### PR TITLE
Feature/6.4/tdi 37602

### PR DESCRIPTION
**Pull request by TotalP**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows Talend standard
- [x] Tests for the changes have been added (for bug fixes / features)

**What is the current behavior?** (You can also link to an open issue here)
There is no "Guess schema" feature

**What is the new behavior?**
"Guess schema" feature added.

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:
https://jira.talendforge.org/browse/TDI-37602

### CODE REVIEW

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/976/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/976/)

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/983/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/983/)

File: components/components-salesforce/src/main/java/org/talend/components/salesforce/runtime/SalesforceSourceOrSink.java [hunk 1]
```diff
@@ -73,6 +76,8 @@
 
     protected static final String SESSION_FILE_PREFX = "sessionIDFile_";
 
+    private static final String EXCEPTION_MESSAGE = "Failed!";
```
Comment (chmyga) [inline components/components-salesforce/src/main/java/org/talend/components/salesforce/runtime/SalesforceSourceOrSink.java]: Is this used at all? if not, maybe it's better to remove this String?

File: components/components-salesforce/src/main/java/org/talend/components/salesforce/runtime/SalesforceSourceOrSink.java [hunk 2]
```diff
@@ -498,4 +503,32 @@ protected boolean isReuseSession() {
...
+    public Schema guessSchema(String soqlQuery) throws IOException, ConnectionException {
+        SoqlQuery query = SoqlQuery.getInstance();
+        query.init(soqlQuery);
+
+        List<FieldDescription> fieldDescriptions = query.getFieldDescriptions();
+        String drivingEntityName = query.getDrivingEntityName();
+
+        SchemaBuilder.FieldAssembler fieldAssembler = SchemaBuilder.record("GuessedSchema").fields();
+
+        DescribeSObjectResult describeSObjectResult = connect(null).connection.describeSObject(drivingEntityName);
```
Comment (chmyga) [inline components/components-salesforce/src/main/java/org/talend/components/salesforce/runtime/SalesforceSourceOrSink.java]: Why don't we use the RuntimeContainer here? if we have one and the connection is already initialized we won't use the existing connection, but will create one.

File: components/components-salesforce/src/main/java/org/talend/components/salesforce/soql/SoqlQuery.java [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+        private String buildSubqueryFullName(String entityName, FieldContext field) {
+            String fieldName = field.getText();
+            String[] splittedFieldNames = fieldName.split("\\.");
+            StringBuilder sb = new StringBuilder();
+            sb.append(entityName);
+            sb.append(RECORDS);
+            for (int i = 0; i < splittedFieldNames.length; i++) {
+                sb.append(splittedFieldNames[i]);
+                sb.append("_");
```
Comment (chmyga) [inline components/components-salesforce/src/main/java/org/talend/components/salesforce/soql/SoqlQuery.java]: Is it possible to use other separators?(comment from Ivan Gonchar)

File: components/components-salesforce/src/main/java/org/talend/components/salesforce/tsalesforceinput/TSalesforceInputProperties.java [hunk 4]
```diff
@@ -73,7 +85,11 @@ public void setupLayout() {
         mainForm.addRow(queryMode);
         mainForm.addRow(condition);
         mainForm.addRow(manualQuery);
+
+        mainForm.addColumn(Widget.widget(guessSchema).setWidgetType(Widget.BUTTON_WIDGET_TYPE));
+
         mainForm.addRow(Widget.widget(query).setWidgetType(Widget.TEXT_AREA_WIDGET_TYPE));
+        query.setValue("\"SELECT Id, Name, IsDeleted FROM Account\"");
```
Comment (chmyga) [inline components/components-salesforce/src/main/java/org/talend/components/salesforce/tsalesforceinput/TSalesforceInputProperties.java]: Better to move it to setupProperties() method, in order to be called only once on properties creation.

**Comment by gonchar-ivan** [comment]
Build was failed because of other reason

**Pull request merged**